File: critter/core/src/main/java/dev/morphia/critter/parser/PropertyFinder.java
package dev.morphia.critter.parser;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import dev.morphia.critter.CritterClassLoader;
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator;
import dev.morphia.critter.parser.gizmo.PropertyModelGenerator;
import dev.morphia.critter.parser.java.CritterParser;
import dev.morphia.mapping.Mapper;

import org.objectweb.asm.ClassReader;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.AnnotationNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldNode;
import org.objectweb.asm.tree.MethodNode;

public class PropertyFinder {
    private final Map<Class<?>, Object> providerMap;
    private final CritterClassLoader classLoader;
    private final boolean runtimeMode;

    public PropertyFinder(Mapper mapper, CritterClassLoader classLoader, boolean runtimeMode) {
        this.providerMap = new LinkedHashMap<>();
        for (var provider : mapper.getConfig().propertyAnnotationProviders()) {
            providerMap.put(provider.provides(), provider);
        }
        this.classLoader = classLoader;
        this.runtimeMode = runtimeMode;
    }

    public List<PropertyModelGenerator> find(Class<?> entityType, ClassNode classNode) {
        List<PropertyModelGenerator> models = new ArrayList<>();
        List<MethodNode> methods = discoverPropertyMethods(classNode);
        if (methods.isEmpty()) {
            List<FieldNode> fields = discoverAllFields(entityType, classNode);
            if (!runtimeMode) {
                classLoader.register(entityType.getName(), CritterGizmoGenerator.INSTANCE.fieldAccessors(entityType, fields));
            }
            for (FieldNode field : fields) {
                if (runtimeMode) {
                    CritterGizmoGenerator.INSTANCE.varHandleAccessor(entityType, classLoader, field);
                } else {
                    CritterGizmoGenerator.INSTANCE.propertyAccessor(entityType, classLoader, field);
                }
                models.add(CritterGizmoGenerator.INSTANCE.propertyModelGenerator(entityType, classLoader, field));
            }
        } else {
            if (!runtimeMode) {
                classLoader.register(entityType.getName(), CritterGizmoGenerator.INSTANCE.methodAccessors(entityType, methods));
            }
            for (MethodNode method : methods) {
                if (runtimeMode) {
                    CritterGizmoGenerator.INSTANCE.varHandleAccessor(entityType, classLoader, method);
                } else {
                    CritterGizmoGenerator.INSTANCE.propertyAccessor(entityType, classLoader, method);
                }
                models.add(CritterGizmoGenerator.INSTANCE.propertyModelGenerator(entityType, classLoader, method));
            }
        }
        return models;
    }

    private boolean isPropertyAnnotated(List<AnnotationNode> annotationNodes, boolean allowUnannotated) {
        List<AnnotationNode> annotations = annotationNodes != null ? annotationNodes : List.of();
        List<String> keys = providerMap.keySet().stream()
                .map(type -> Type.getType(type).getDescriptor())
                .toList();
        return allowUnannotated || annotations.stream().anyMatch(a -> keys.contains(a.desc));
    }

    private List<FieldNode> discoverAllFields(Class<?> entityType, ClassNode classNode) {
        List<FieldNode> fields = new ArrayList<>();
        Map<String, Boolean> seen = new LinkedHashMap<>();
        Class<?> current = entityType;
        ClassNode currentNode = classNode;

        while (current != null && current != Object.class) {
            ClassNode node = currentNode != null ? currentNode : readClassNode(current);
            if (node == null)
                break;
            for (FieldNode field : discoverFields(node)) {
                if (seen.putIfAbsent(field.name, Boolean.TRUE) == null) {
                    fields.add(field);
                }
            }
            current = current.getSuperclass();
            currentNode = null;
        }
        return fields;
    }

    private ClassNode readClassNode(Class<?> type) {
        String resourceName = type.getName().replace('.', '/') + ".class";
        InputStream inputStream = type.getClassLoader().getResourceAsStream(resourceName);
        if (inputStream == null)
            return null;
        ClassNode node = new ClassNode();
        try {
            new ClassReader(inputStream).accept(node, 0);
        } catch (IOException e) {
            return null;
        }
        return node;
    }

    private List<FieldNode> discoverFields(ClassNode classNode) {
        List<String> transientDescs = CritterParser.INSTANCE.transientAnnotations();
        List<FieldNode> result = new ArrayList<>();
        for (FieldNode field : classNode.fields) {
            List<AnnotationNode> visible = field.visibleAnnotations != null ? field.visibleAnnotations : List.of();
            boolean isTransient = visible.stream().map(a -> a.desc).anyMatch(transientDescs::contains);
            if (!isTransient && isPropertyAnnotated(field.visibleAnnotations, true)) {
                result.add(field);
            }
        }
        return result;
    }

    private List<MethodNode> discoverPropertyMethods(ClassNode classNode) {
        List<MethodNode> result = new ArrayList<>();
        for (MethodNode method : classNode.methods) {
            if (method.name.startsWith("get") && Type.getArgumentTypes(method.desc).length == 0) {
                if (isPropertyAnnotated(method.visibleAnnotations, false)) {
                    result.add(method);
                }
            }
        }
        return result;
    }
}

File: critter/core/src/main/kotlin/dev/morphia/critter/parser/PropertyFinder.kt
package dev.morphia.critter.parser

import dev.morphia.critter.CritterClassLoader
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator.accessor
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator.fieldAccessors
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator.methodAccessors
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator.propertyModelGenerator
import dev.morphia.critter.parser.gizmo.PropertyModelGenerator
import dev.morphia.critter.parser.java.CritterParser
import dev.morphia.mapping.Mapper
import org.objectweb.asm.ClassReader
import org.objectweb.asm.Type
import org.objectweb.asm.tree.AnnotationNode
import org.objectweb.asm.tree.ClassNode
import org.objectweb.asm.tree.FieldNode

class PropertyFinder(mapper: Mapper, val classLoader: CritterClassLoader) {
    private val providerMap =
        mapper.config.propertyAnnotationProviders().associateBy { it.provides() }

    fun find(entityType: Class<*>, classNode: ClassNode): List<PropertyModelGenerator> {
        val models = mutableListOf<PropertyModelGenerator>()
        val methods = discoverPropertyMethods(classNode)
        if (methods.isEmpty()) {
            val fields = discoverAllFields(entityType, classNode)
            classLoader.register(entityType.name, fieldAccessors(entityType, fields))
            fields.forEach { field ->
                accessor(entityType, classLoader, field)
                models += propertyModelGenerator(entityType, classLoader, field)
            }
        } else {
            classLoader.register(entityType.name, methodAccessors(entityType, methods))
            methods.forEach { method ->
                accessor(entityType, classLoader, method)
                models += propertyModelGenerator(entityType, classLoader, method)
            }
        }

        return models
    }

    private fun isPropertyAnnotated(
        annotationNodes: MutableList<AnnotationNode>?,
        allowUnannotated: Boolean,
    ): Boolean {
        val annotations = annotationNodes ?: listOf()
        val keys = providerMap.keys.map { Type.getType(it).descriptor }
        return allowUnannotated || annotations.any { a -> a.desc in keys }
    }

    private fun discoverAllFields(entityType: Class<*>, classNode: ClassNode): List<FieldNode> {
        val fields = mutableListOf<FieldNode>()
        val seen = mutableSetOf<String>()
        var current: Class<*>? = entityType
        var currentNode: ClassNode? = classNode

        while (current != null && current != Object::class.java) {
            val node = currentNode ?: readClassNode(current) ?: break
            discoverFields(node).filter { seen.add(it.name) }.let { fields.addAll(it) }
            current = current.superclass
            currentNode = null
        }
        return fields
    }

    private fun readClassNode(type: Class<*>): ClassNode? {
        val resourceName = type.name.replace('.', '/') + ".class"
        val inputStream = type.classLoader.getResourceAsStream(resourceName) ?: return null
        val node = ClassNode()
        ClassReader(inputStream).accept(node, 0)
        return node
    }

    private fun discoverFields(classNode: ClassNode) =
        classNode.fields
            .filter { field ->
                (field.visibleAnnotations ?: listOf())
                    .map { a -> a.desc }
                    .none { desc -> desc in CritterParser.transientAnnotations() }
            }
            .filter { isPropertyAnnotated(it.visibleAnnotations, true) }

    private fun discoverPropertyMethods(classNode: ClassNode) =
        classNode.methods
            .filter { it.name.startsWith("get") && Type.getArgumentTypes(it.desc).isEmpty() }
            .filter { isPropertyAnnotated(it.visibleAnnotations, false) }
}
